DIR857GPL/SRC/ubicom-distro-3.0/uClinux/openswan-2.6.24rc4/programs/pluto/ipsec.secrets.5
'\" t
.\"     Title: IPSEC.SECRETS
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.75.1 <http://docbook.sf.net/>
.\"      Date: 06/29/2010
.\"    Manual: [FIXME: manual]
.\"    Source: [FIXME: source]
.\"  Language: English
.\"
.TH "IPSEC\&.SECRETS" "5" "06/29/2010" "[FIXME: source]" "[FIXME: manual]"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ipsec.secrets \- secrets for IKE/IPsec authentication
.SH "DESCRIPTION"
.PP
The file
ipsec\&.secrets
contains a list of secrets, aka preshared secrets, RSA signatures, or pointers to X\&.509 Digital Certificates\&. These secrets are used by
\fBipsec_pluto\fR(8)
, the Openswan Internet Key Exchange daemon, to authenticate other hosts\&. Currently there are five kinds of secrets: preshared secrets, RSA private keys, passphrases for X\&.509 certificates and if compiled with
\fBUSE_XAUTH=true\fR
there is support for XAUTH static passwords\&. Smartcard support has been moved to the NSS framework\&.
.PP
It is vital that these secrets be protected\&. The file should be owned by root, and permissions should be set to block all access by others\&. (eg: chmod 600)
.PP
The file is a sequence of entries and include directives\&. Here is an example \- each entry or directive must start at the left margin, but if it continues beyond a single line, each continuation line must be indented\&.
.sp
.if n \{\
.RS 4
.\}
.nf
# sample /etc/ipsec\&.secrets file for 10\&.1\&.0\&.1
10\&.1\&.0\&.1 10\&.2\&.0\&.1: PSK "secret shared by two hosts"
# sample roadwarrior
%any gateway\&.corp\&.com: PSK "shared secret with many roadwarriors"
# sample server for roadwarriors
myip %any : PSK "shared secret with many roadwarriors"

# an entry may be split across lines,
# but indentation matters
www\&.xs4all\&.nl @www\&.kremvax\&.ru
\ \&\ \&\ \&\ \&10\&.6\&.0\&.1 10\&.7\&.0\&.1 1\&.8\&.0\&.1: PSK "secret shared by 5 systems"

# an RSA private key\&.
# note that the lines are too wide for a
# man page, so \&.\&.\&. has been substituted for
# the truncated part
@my\&.com: rsa {
\ \&\ \&\ \&\ \&Modulus:\ \&0syXpo/6waam+ZhSs8Lt6jnBzu3C4grtt\&.\&.\&.
\ \&\ \&\ \&\ \&PublicExponent:\ \&0sAw==
\ \&\ \&\ \&\ \&PrivateExponent:\ \&0shlGbVR1m8Z+7rhzSyenCaBN\&.\&.\&.
\ \&\ \&\ \&\ \&Prime1:\ \&0s8njV7WTxzVzRz7AP+0OraDxmEAt1BL5l\&.\&.\&.
\ \&\ \&\ \&\ \&Prime2:\ \&0s1LgR7/oUMo9BvfU8yRFNos1s211KX5K0\&.\&.\&.
\ \&\ \&\ \&\ \&Exponent1:\ \&0soaXj85ihM5M2inVf/NfHmtLutVz4r\&.\&.\&.
\ \&\ \&\ \&\ \&Exponent2:\ \&0sjdAL9VFizF+BKU4ohguJFzOd55OG6\&.\&.\&.
\ \&\ \&\ \&\ \&Coefficient:\ \&0sK1LWwgnNrNFGZsS/2GuMBg9nYVZ\&.\&.\&.
\ \&\ \&\ \&\ \&}

# An X\&.509 pem encoded private key file with (optional) passphrase 
: RSA vpnserverKey\&.pem "<optional passphrase>"
# An X\&.509 pem encoded private key file locked with a passphrase
# Note: the %prompt keyword means someone has to actually enter the passphrase
# at load time \- usually via \fBipsec_whack\fR(8)
:  RSA vpnserverKey\&.pem %prompt

# XAUTH password, used with leftxauthusername=username
@username : XAUTH "password"

include ipsec\&.*\&.secrets	# get secrets from other files
.fi
.if n \{\
.RE
.\}
.sp

Each entry in the file is a list of indices, followed by a secret\&. The two parts are separated by a colon (\fB:\fR) that is followed by whitespace or a newline\&. For compatibility with the previous form of this file, if the key part is just a double\-quoted string the colon may be left out\&. If filenames are not absolute paths, they are relative to the
\fBipsec\&.d/private/\fR
directory\&.
.PP
An index is an IP address, or a Fully Qualified Domain Name, user@FQDN,
\fB%any\fR
or
\fB%any6\fR
(other kinds may come)\&. An IP address may be written in the familiar dotted quad form or as a domain name to be looked up when the file is loaded (or in any of the forms supported by the Openswan
\fBipsec_ttoaddr\fR(3)
routine)\&. In many cases it is a bad idea to use domain names because the name server may not be running or may be insecure\&. To denote a Fully Qualified Domain Name (as opposed to an IP address denoted by its domain name), precede the name with an at sign (\fB@\fR)\&.
.PP
Matching IDs with indices is fairly straightforward: they have to be equal\&. In the case of a \(lqRoad Warrior\(rq connection, if an equal match is not found for the Peer\'s ID, and it is in the form of an IP address, an index of
\fB%any\fR
will match the peer\'s IP address if IPV4 and
\fB%any6\fR
will match a the peer\'s IP address if IPV6\&. Currently, the obsolete notation
0\&.0\&.0\&.0
may be used in place of
\fB%any\fR, but please stop doing this, as it will likely stop working around Openswan v3\&.0\&.
.PP
This file is only read at startup time\&. If any changes are made to this file, the pluto daemon should be told to re\-read this file using the command
\fBipsec secrets\fR
or
\fBipsec auto \-\-rereadsecrets\fR\&. If there are any keyfiles protected by a passphrase using
\fB%prompt\fR, you will be prompted again to enter these passphrases\&. To skip the prompting, just hit return to skip unlocking that particular private key\&. Note that currently there is no way to add a specific new entry \- it\'s all or nothing\&.
.PP
Smartcard support has been moved from Openswan to NSS\&. Please see the NSS documentation on how to configure smartcards\&.
.PP
An additional complexity arises in the case of authentication by preshared secret: the responder will need to look up the secret before the Peer\'s ID payload has been decoded, so the ID used will be the IP address\&.
.PP
To authenticate a connection between two hosts, the entry that most specifically matches the host and peer IDs is used\&. An entry with no index will match any host and peer\&. More specifically, an entry with one index will match a host and peer if the index matches the host\'s ID (the peer isn\'t considered)\&. Still more specifically, an entry with multiple indices will match a host and peer if the host ID and peer ID each match one of the indices\&. If the key is for an asymmetric authentication technique (i\&.e\&. a public key system such as RSA), an entry with multiple indices will match a host and peer even if only the host ID matches an index (it is presumed that the multiple indices are all identities of the host)\&. It is acceptable for two entries to be the best match as long as they agree about the secret or private key\&.
.PP
Authentication by preshared secret requires that both systems find the identical secret (the secret is not actually transmitted by the IKE protocol)\&. If both the host and peer appear in the index list, the same entry will be suitable for both systems so verbatim copying between systems can be used\&. This naturally extends to larger groups sharing the same secret\&. Thus multiple\-index entries are best for PSK authentication\&.
.PP
Authentication by RSA Signatures requires that each host have its own private key\&. A host could reasonably use a different private keys for different interfaces and for different peers\&. But it would not be normal to share entries between systems\&. Thus no\-index and one\-index forms of entry often make sense for RSA Signature authentication\&.
.PP
The key part of an entry may start with a token indicating the kind of key\&. \(lqRSA\(rq signifies RSA private key and \(lqPSK\(rq signifies PreShared Key (case is ignored)\&. For compatability with previous forms of this file, PSK is the default\&.
.PP
The token \(lqXAUTH\(rq indicates a eXtended Authentication password\&. There should be one indice, and it should be in the @FQDN format\&. The file will be searched with the XAUTH username, which is usually provided in the configuration file\&. XAUTH is otherwise identical to PSK in syntax\&.
.PP
If the RSA points to a filename, this is assumed to be a PEM (or DER?) encoded X\&.509 private key\&. The private key may be protected by a 3DES encryption\&. 1DES encrypted key files will be rejected\&. If the private key is protected by a passphrase and this passphrase is not specified in
\fBipsec\&.secrets\fR, the connection cannot be automatically started using
\fBauto=start\fR, but instead must be brought up using
\fBipsec auto \-\-up connname\fR, upon which the user will be prompted for the passphrase to unlock the private key belonging to the X\&.509 certificate\&. PKCS#12 files, which include the private key file, cannot be specified in
\fBipsec\&.secrets\fR\&. Private keys can be extracted from PKCS#12 files using the following command:
\fBopenssl pkcs12 \-nocerts \-in clientCert\&.p12 \-out clientKey\&.pem\fR
.PP
A preshared secret is most conveniently represented as a sequence of characters, delimited by the double\-quote character (\fB"\fR)\&. The sequence cannot contain a newline or double\-quote\&. Strictly speaking, the secret is actually the sequence of bytes that is used in the file to represent the sequence of characters (excluding the delimiters)\&. A preshared secret may also be represented, without quotes, in any form supported by
\fBipsec_ttodata\fR(3)\&.
.PP
An RSA private key is a composite of eight generally large numbers\&. The notation used is a brace\-enclosed list of field name and value pairs (see the example above)\&. A suitable key, in a suitable format, may be generated by
\fBipsec_rsasigkey\fR(8)\&. The structure is very similar to that used by BIND 8\&.2\&.2 or later, but note that the numbers must have a \(lq0s\(rq prefix if they are in base 64\&. The order of the fields is fixed\&.
.PP
The first token an entry must start in the first column of its line\&. Subsequent tokens must be separated by whitespace, except for a colon token, which only needs to be followed by whitespace\&. A newline is taken as whitespace, but every line of an entry after the first must be indented\&.
.PP
Whitespace at the end of a line is ignored (except in the 0t notation for a key)\&. At the start of line or after whitespace,
\fB#\fR
and the following text up to the end of the line is treated as a comment\&. Within entries, all lines must be indented (except for lines with no tokens)\&. Outside entries, no line may be indented (this is to make sure that the file layout reflects its structure)\&.
.PP
An include directive causes the contents of the named file to be processed before continuing with the current file\&. The filename is subject to \(lqglobbing\(rq as in
\fBsh\fR(1), so every file with a matching name is processed\&. Includes may be nested to a modest depth (10, currently)\&. If the filename doesn\'t start with a
\fB/\fR, the directory containing the current file is prepended to the name\&. The include directive is a line that starts with the word
\fBinclude\fR, followed by whitespace, followed by the filename (which must not contain whitespace)\&.
.SH "FILES"
.PP
/etc/ipsec\&.secrets
.SH "SEE ALSO"
.PP
The rest of the Openswan distribution, in particular
\fBipsec.conf\fR(5),
\fBipsec\fR(8),
\fBipsec_newhostkey\fR(8),
\fBipsec_rsasigkey\fR(8),
\fBipsec_showhostkey\fR(8),
\fBipsec_auto\fR(8)
\fB\-\-rereadsecrets\fR, and
\fBipsec_pluto\fR(8)
\fB\-\-listen\fR,\&.
BIND 8\&.2\&.2 or later,
\m[blue]\fBftp://ftp\&.isc\&.org/isc/bind/src/\fR\m[]
.SH "HISTORY"
.PP
Originally designed for the FreeS/WAN project <\m[blue]\fBhttp://www\&.freeswan\&.org\fR\m[]> by D\&. Hugh Redelmeier\&. Updated for Openswan by Ken Bantoft\&.
.SH "BUGS"
.PP
If an ID is
0\&.0\&.0\&.0, it will match
\fB%any\fR; if it is
\fB0::0\fR, it will match
\fB%any6\fR\&.


DIR657GPL/SRC/ubicom-distro/uClinux/user/cmds/security/vpn/ipsec.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <syslog.h>
#include <ctype.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/ioctl.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <net/if.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <net/if_arp.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <strings.h>
#include <string.h>

#include "cmds.h"
#include "shutils.h"
#include <nvramcmd.h>
#define uint32 unsigned long
#define uint unsigned int
#define uint8 unsigned char
#define uint64 unsigned long long

#define IPSEC_CONF_FILE		"/tmp/ipsec.conf"
#define IPSEC_SECRETS_FILE	"/tmp/ipsec.secrets"

char *info_getwan_interface()
{
	char *wan_proto = nvram_safe_get("wan_proto");

	if (strcmp(wan_proto, "static") == 0 || strcmp(wan_proto, "dhcpc") == 0)
		return "eth0.1";
	return "ppp0";
}

static char *strsep_with_validation(char **, char *, char *);

#define MAX_NVRAM_IPSEC_VALUE_SIZE 1024
#define STRSEP strsep_with_validation
#include "ipsec_bcm.h"
static char * strsep_with_validation(char **ptr,char *sep,char * retchar)	
{ 
	char * strptr ;	
	char *cptr;

	if (*ptr == NULL)
		return retchar;	

	strptr  =strsep(ptr,sep);	
	if (strptr != NULL) {

		/* change eny last \n to NULL */	
		cptr = strchr(strptr,'\r');;
		if (cptr != NULL)
			*cptr =' '; 


		if (*strptr == 0x00)		
			return retchar;	
	
	}else if (strptr == NULL)
		return retchar;
	
	return	strptr;	
}

static int ipsec_update_wan_addr(char *buf, const char *ip)
{
	char *t, *n, *p;
	int rev = -1;

	//cprintf("update_wan_addr:%s\n", ip);
	t = buf;
	if ((t = strchr(t, ';')) == NULL) goto out; t++; /* skip action field */
	if ((t = strchr(t, ';')) == NULL) goto out; t++; /* skip name field */
	if ((t = strchr(t, ',')) == NULL) goto out; t++; /* skip tunnel field*/
	if ((n = strchr(t, '-')) == NULL) goto out;
	//cprintf("Updating:%s\n%s\n", t, n);
	p = strdup(n);
	strcpy(t, ip); /* It would overwrite bufffer @buf field of Left */
	strcat(t, p);
	free(p);
	rev = 0;
	//cprintf("BUF[%s]\n", buf);
out:
	return rev;
}

/* replace 'key' as 'rep' from 'org' string */
static void
string_replace(char *org, const char *key, const char *rep)
{
	char buf[512];
	char *idx = strstr(org, key);

	if (idx == NULL)
		return;

	bzero(buf, sizeof(buf));
	strncpy(buf, org, idx - org);
	strcat(buf, rep);
	strcat(buf, idx + strlen(key));
	strcpy(org, buf);
}
static FILE *get_and_clear_ipsec_conf(const char *secrets, const char *conf)
{
	FILE *fd;

	/* clear ipsec.secrets); */
	if (!(fd = fopen(secrets, "w"))) {
		perror(secrets);
		return NULL;
	}
	fclose(fd);

	/* Write to ipsec.conf  file based on current information */
	if (!(fd = fopen(conf, "w"))) {
		perror(conf);
		return NULL;
	}
	return fd;
}


static char *strcat_ipsec_string(int vpn_idx, char *buf)
{
	char vpn_conn[] = "vpn_connXXXXXX";
	char vpn_extra[] = "vpn_extraXXXXXX";
	char *vp;
	sprintf(vpn_conn, "vpn_conn%d", vpn_idx);
	sprintf(vpn_extra, "vpn_extra%d", vpn_idx);

	if (*(vp = nvram_safe_get(vpn_conn)) == '\0')
		return NULL;
	/* Skip disabled rule */
	if ((strncmp(vp, "ignore", 6)) == 0)
		return NULL;
	
	strcpy(buf, vp);
	strcat(buf, ";");
	if (*(vp = nvram_safe_get(vpn_extra)) != "")
		strcat(buf, vp);
	return buf;

}
/*
* create/recreate the ipsec.conf file in /etc
*/
int
create_ipsec_conf(void)
{
	FILE *fd;
	char *wan_ifname=NULL;
	char *lan_ifname=NULL;
	char *wl_ifname=NULL;
	char wanif0[IF_NAMESIZE] = "";
	char wanif1[IF_NAMESIZE] = "";
	char host0[32] = "", host1[32] = "", mask0[32] = "", mask1[32] = "";
	
	//ipsec_Conn_t ipsec_Conn;
	int conn_id=1; /* connections start from 1. not 0 */

	fd = get_and_clear_ipsec_conf(IPSEC_SECRETS_FILE, IPSEC_CONF_FILE);

#ifdef CONFIG_BCM_IPSEC
	lan_ifname= nvram_safe_get("lan_ifname");
	wl_ifname= nvram_safe_get("wl_ifname");
	/* DIR130/330 style */
	dual_wans_info(host0, mask0, wanif0, host1, mask1, wanif1);
#else
	/* FIXME: hard code in DIR730*/
	lan_ifname= "br0";
	wl_ifname= "eth1";
#endif	
#ifdef CONFIG_BCM_IPSEC
	if (*wanif0 != 0x00 && *wanif1 != 0x00) {
		wan_ifname = wanif0;
		/* XXX: If daul wans up. overwirte wireless IFNAME
		 * After all, Ipses does not bind wireless.
		 * */
		wl_ifname = wanif1;
	} else if (*wanif0 != 0x00) {
		wan_ifname = wanif0;
	} else if (*wanif1 != 0x00)  {
		wan_ifname = wanif1;
	} else {
		fprintf(stderr, "IPSec: interfaces error!\n");
		return -1;
	}
#else
	/* FIXME: bad hard code */
	strcpy(wanif0, info_getwan_interface());
	get_ip(wanif0, host0);
#endif
	fprintf(fd, "config setup\n");

#ifdef CONFIG_BCM_IPSEC
	__ipsec_bcm_setup(fd, wan_ifname, lan_ifname, wl_ifname);
	if (strcmp(nvram_safe_get("vpn_nat_traversal"), "1") == 0 ) {
		fputs("\tnat_traversal=yes\n", fd);
		fputs("\tvirtual_private=%v4:0.0.0.0/1,%v:128.0.0.0/1,%all\n", fd);
	}

	/* default for subsequent connections */
	fputs("\nconn %default\n\
		\tkeyingtries=0\n\
		\tdisablearrivalcheck=no\n\
		\tauthby=rsasig\n\
		\tleftrsasigkey=%dnsondemand\n\
		\trightrsasigkey=%dnsondemand\n",fd);
#else
	fprintf(fd,"\tnat_traversal=yes\n"
	"\tvirtual_private=%v4:10.0.0.0/8,%v4:192.168.0.0/16,%4:172.16.0.0/12\n"
	"\tprotostack=netkey\n\n");
						
#endif
	conn_id=1;
	{
#define VPN_BUFFER_SIZE	1024
#define VPN_CONN_MAX	26   //99//26
		FILE *authfd;
		char *vp, buf[VPN_BUFFER_SIZE];
		int vpn_idx;
		
		if (!(authfd = fopen(IPSEC_SECRETS_FILE, "a+"))) {
			perror(IPSEC_CONF_FILE);
			return -1;
		};		
		//init_ipsec_script(); /* obsolete */
		for (vpn_idx = 0; vpn_idx <= VPN_CONN_MAX; vpn_idx++) {
			if ((vp = strcat_ipsec_string(vpn_idx, buf)) == NULL)
				continue;
			fprintf(stderr, "wanif0: %s, wanif1: %s\n", wanif0, wanif1);
			if (*wanif0 != 0x00 || (*wanif0 == 0x00 && (NVRAM_MATCH("wan0_proto", "rupptp") ||
				NVRAM_MATCH("wan0_proto", "rupppoe")))) {
				char route[8], uname[16], tunnel[16];
				char range[64];
				char leftip[32], rightip[256];

				bzero(route, sizeof(route));
				bzero(uname, sizeof(uname));
				bzero(tunnel, sizeof(tunnel));
				bzero(range, sizeof(range));
				bzero(leftip, sizeof(leftip));
				bzero(rightip, sizeof(rightip));
				/* route;s2s;tunnel,0.0.0.0-172.21.33.23,192.168.... */
				/* add;ruser;tunnel,10.0.0.2-any%,192.168.... */
				sscanf(vp, "%[^;];%[^;];%[^,],%[^,]", route, uname, tunnel, range);
				sscanf(range, "%[^-]-%s", leftip, rightip);
				if (*wanif1 != 0x00) {	// dual access
                                        /* XXX: DUAL ACCESS 2010/4/21 copy from DIR130, not verify */
					if (strcmp(rightip, "%any") == 0) {	// remote user
						char buf2[VPN_BUFFER_SIZE];
						char comm_name1[VPN_BUFFER_SIZE];
						char comm_name2[VPN_BUFFER_SIZE];

						sscanf(buf, "%[^;];%[^;]", route, uname);
						bzero(buf2, sizeof(buf2));
						bzero(comm_name1, sizeof(comm_name1));
						bzero(comm_name2, sizeof(comm_name2));

						sprintf(comm_name1, "%s_russia_remote_ppp", uname);
						sprintf(comm_name2, "%s_russia_remote_phy", uname);
						strncpy(buf2, buf, sizeof(buf2));
						//cprintf("original buf: %s\n", buf);
						//cprintf("original buf2: %s\n", buf2);
						string_replace(buf, uname, comm_name1);
						string_replace(buf2, uname, comm_name2);

						//cprintf("uname: %s\n", uname);
						cprintf("comm_name1: %s\n", comm_name1);
						cprintf("comm_name2: %s\n", comm_name2);
						//cprintf("processed buf: %s\n", buf);
						//cprintf("processed buf2: %s\n", buf2);
						cprintf("(dual access remote user)Init IPSec: %s:%s\n", wanif0, host0);
						cprintf("(dual access remote user)Init IPSec: %s:%s\n", wanif1, host1);
						ipsec_update_wan_addr(buf, host0);
						if (conf_vpn_connection(buf, fd, authfd) == 0)
							conn_id++;

						ipsec_update_wan_addr(buf2, host1);
						if (conf_vpn_connection(buf2, fd, authfd) == 0)
							conn_id++;
                                                /***********END DUAL ACCESS***************/

					} else {	// site to site
						char dev[8], dev_gw[16], dev_ip[16];

						//getGatewayInterfaceByIP(host1, dev, dev_gw);
						bzero(dev, sizeof(dev));
						bzero(dev_gw, sizeof(dev_gw));
						bzero(dev_ip, sizeof(dev_ip));

						sleep(3);
						getGatewayInterfaceByIP(rightip, dev, dev_gw);
						getIPbyDevice(dev, dev_ip);

						if (strlen(dev) == 0 || strlen(dev_ip) == 0)
							continue;

						cprintf("(dual access site to site)Init IPSec: %s:%s\n", dev, dev_ip);
						ipsec_update_wan_addr(buf, dev_ip);
						if (conf_vpn_connection(buf, fd, authfd) == 0)
							conn_id++;
					}
				} else { // single access
					cprintf("Init IPSec: %s:%s,[%s]\n", wanif0, host0, buf);
					ipsec_update_wan_addr(buf, host0);
					if (conf_vpn_connection(buf, fd, authfd) == 0)
						conn_id++;
				}
			}
		}
		fclose(authfd);
	}
	fclose(fd);
	return (conn_id-1);
}
#ifndef CONFIG_BCM_IPSEC
/*
 * Start IPSec.
 */
int start_ipsec(void)
{
	int noofcons=0;
	noofcons =create_ipsec_conf();
#ifndef CONFIG_UBICOM_ARCH
	if (noofcons > 0) {
		eval("ipsec", "setup", "start");
		return 0;
	}
#endif //CONFIG_UBICOM_ARCH
	/* no tunnel profile */
	return -1;
}

/*
 * Stop IPSec.
 */
static int stop_ipsec(void)
{
#ifdef CONFIG_UBICOM_ARCH
	config_whack_stop_pluto();
	eval("killall", "-9", "ipsec_dns");
#else
	eval("ipsec", "setup", "stop");
#endif
	return 0;
}
#endif //CONFIG_BCM_IPSEC
/*
 * Returns an IP address as an unsinged long 
 */
inline unsigned long ipaddr2hex(char *str)
{
	int d1,d2,d3,d4;
	char *s;

	s = str; d1 = atoi (s);
	s = strchr(s + 1, '.'); d2 = atoi (s+1);
	s = strchr(s + 1, '.'); d3 = atoi (s+1);
	s = strchr(s + 1, '.'); d4 = atoi (s+1);
	return ((unsigned long) d1 << 24 | (unsigned long)d2 << 16 
			| (unsigned long) d3 <<8 | (unsigned long) d4 );
}

/*
 * Translate ip mask into bits 
 */
static int mask2bits (char *str)
{
	unsigned long int x;
	int bits = 0;

	x = ipaddr2hex(str);
	while (!(x & 1) && (bits < 31)) { bits++; x >>= 1; }

	return 32 - bits;
}

extern int hw_ipsec_clear();
static int ipsec_main_start(int argc, char *argv[])
{
	if (start_ipsec() == -1)
		return -1;
	//system("iptables -t nat -D POSTROUTING -o eth1 -j MASQUERADE >/dev/null 2>&1");
	//system("iptables -t nat -I POSTROUTING -o eth1 -j MASQUERADE >/dev/null 2>&1");
	//system("echo \"1\" > /proc/sys/net/ipv4/ip_forward");
	//hw_ipsec_clear();
	/* FIXME: if IPSec ready, delete this rule. This rule be added from
	 * rc start_firewall, because of to prevent src IP is DUT LAN. But
	 * this is not good idea, conntrack problem still have.
	 *
	 * FredHung, 2010/6/10
	 * */

	unlink("/tmp/ns.res");
	if (fork() == 0) {
		sleep(10);
		system("iptables -t nat -D POSTROUTING -o ipsec0 -j MASQUERADE");
		exit(-1);
	}

	if (access("/tmp/ns.res", F_OK) == 0)
                system("ipsec_dns /tmp/ns.res &");

	return 0;
}

static int ipsec_main_stop(int argc, char *argv[])
{
	system("iptables -t nat -D POSTROUTING -o eth1 -j MASQUERADE");
	unlink("/tmp/ns.res");
	stop_ipsec();
}

extern int ipsec_main_monitor(int argc, char *argv[]);
int ipsec_main(int argc, char *argv[])
{
	struct subcmd cmds[] = {
		{ "start", ipsec_main_start, "start IPSec: rev0.00"},
		{ "stop", ipsec_main_stop, "stop IPSec: rev0.00"},
		{ "monitor", ipsec_main_monitor, "ike rekey timer" },
		{ NULL, NULL}
	};
	int rev, lock;
	setenv("PATH", "/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin", 1);
	if (nvram_get("ipsec_lock") == NULL)
		lock = 0;
	else
		lock = 1;
	rd(1, "lock:%d", lock);
	lock_prog(argc, argv, lock);
	rev = lookup_subcmd_then_callout(argc, argv, cmds);
	unlock_prog();
	return rev;
}
